Initial draft of k8s without helm reference (#5)
sourcegraph/customer-training: collateral/deployment/k8s-without-helm-reference.md

# Deploying Sourcegraph to Kubernetes (k8s) without Helm

## Introduction

> This material is intended to be used as a handout to customers working with the Customer Engineering team. The materials are meant to serve as an overview and resource to aid in deep diving into Sourcegraph's documentation and common deployment tasks.

Reviewer: ```suggestion
This material is intended to be used as a handout to customers working with a Sourcegraph Customer Engineer to deploy their own instance. The materials are meant to serve as an overview and resource to aid in deep diving into Sourcegraph's documentation and common deployment tasks.
```

Clarifying here to add SG explicitly + confirm the task at hand, since we do lots of stuff

**Table of Contents:**

- [Introduction](#introduction)
- [Sourcegraph Setup and Deployment](#sourcegraph-setup-and-deployment)
	- [Prerequisites](#prerequisites)
  - [Create your cluster](#create-your-cluster)
  - [Configure your manifest file](#configure-your-manifest-file)
  - [Install Sourcegraph](#install-sourcegraph)
- [Next Steps](#next-steps)

## Sourcegraph Setup and Deployment

### Prerequisites

> - Ensure you have your Sourcegraph Enterprise License

Reviewer: Up to you on whether you want to leave this, but this isn't required to deploy, and suggesting that it is confuses a lot of people historically (typically they receive the license after the instance is stood up.)

```suggestion
```

Author: @emchap are they modifying their manifest and applying again afterward? I guess if I was going to do this, I'd create my cluster and deploy with all my configs in the manifest file then later changes/tweaks to my manifest could be applied later if needed, but a license to get sourcegraph started seemed like a "first deploy" thing versus a later modification.

Totally open to your thoughts since you've gone through implementation here and have the insight.

Reviewer: @IAmHughes The license is added in-app and is unrelated to the deployment; it's usually given to them once the deployment is stood up, and the CE will run through checks with them at this point.

Author: Interesting. @emchap would you say the best practice is to deploy my instance....then apply the manifest again with the license key later? As a best practice, it feels like having my license key initially when I modify my manifest for storage classes, ingress, etc makes more sense than applying it afterward.

I know you're speaking to how it has been, but is that really a best practice in general for standing up software like this? I haven't seen it that way in my past experience, but my past experience is a narrow sliver of the entire software ecosystem of course. Is there a benefit to adding the license later?

- Access to deploy Kubernetes clusters
  - Minimum Kubernetes version: `v1.19` and `kubectl` `v1.19`

>   - Support for Persistent Volumes (SSDs recommended)

Reviewer: ```suggestion
  - Support for Persistent Volumes (SSDs strongly recommended)
```
I might even change this to "required", the delivery team treats them as a prereq.

Reviewer 2: +1 to Emily here, solid state drives are typically created as [required](https://docs.sourcegraph.com/admin/deployment_best_practices#kubernetes)

> - Determine resource requirements to properly size your instance: [Resource Estimator](https://docs.sourcegraph.com/admin/deploy/resource_estimator)

Reviewer: I defer to you on how you want to phrase this, but the resource estimator often underestimates resources by about 40%, so this functionally is really "use the resource estimator as a starting-off point but pick values in conjunction with the CE"

Author: Hmm, this is a good point actually. I think adding something about the CE works for sure, but in general we should have resources that work. If the resource estimator isn't working (which I think I heard elsewhere it isn't always accurate too), that tool should be removed or improved, right? I think there's a team working to improve the estimator but I'm not sure what the stage is.

How does this sound as a "for now" thing and hopefully in the future we can rely more on the resource estimator, or remove it if it doesn't work properly.

```suggestion
- Determine resource requirements to properly size your instance: The [Resource Estimator](https://docs.sourcegraph.com/admin/deploy/resource_estimator) can give you a baseline, however work with your Sourcegraph Customer Engineer to ensure you are sized appropriately
```

Reviewer: Don't disagree! But it is the way it currently is (and has been for at least the last 18 months; the whole thing is a giant 100mb application and probably will just get completely rewritten). I believe Delivery is working on improvements this quarter and you can check with them on status.

Author: > Don't disagree! But it is the way it currently is (and has been for at least the last 18 months; the whole thing is a giant 100mb application and probably will just get completely rewritten). I believe Delivery is working on improvements this quarter and you can check with them on status.

@amenne tagging you just for awareness, @jwayman as well. This tool is in the docs and isn't useful in it's current state, to the extent that we have been calling out that working with a CE is going to be a better route. I'm going to keep the link to the. resource estimator and the suggested change above of adding in CE language, but I wanted to make sure this is elevated again that we're saying here that our official docs aren't sufficient without CE involvement.

- Install relevant CLIs for your provider of choice:
  - AWS: [`aws` cli](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) and [`ejsctk`](https://eksctl.io/introduction/#installation)
  - Azure: [`az` cli](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
  - GCP: [`gcloud` cli](https://cloud.google.com/sdk/gcloud)

### Create your cluster

> - Follow the instructions for your provider of choice to create your Kubernetes (k8s) cluster: [Cloud installation instructions](https://docs.sourcegraph.com/admin/deploy/kubernetes#cloud-installation)

Reviewer: @mike-r-mclaughlin Are the values suggested in this doc good or no?

Author: If they aren't, let's PR the original doc so we have a good single source of truth. 🙏

Reviewer: I think the issue is more that the appropriate nodes to use vary deployment-by-deployment (https://stackoverflow.com/c/sourcegraph/questions/187/188#188 has more context on AWS node choice), but I defer to Mike.

Author: Will totes wait for @mike-r-mclaughlin, thanks!

I do think regardless if the information isn't correct in our official documentation we need a better way to show that in the documentation. This is to reference the official docs, not be a separate documentation piece (or a second source of truth).

  - Follow the recommendations for `node type`, `disk type`, and `minimum` and `maximum` nodes

### Configure your manifest file

> Sourcegraph has a list of configurations available and recommended that need to be made to your manifest file to properly install Sourcegraph into your cluster. Please review the full configuration guide prior to proceeding: [Configure Sourcegraph with Kubernetes](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure)

Reviewer: I worry that this is just pointing them to a very long doc without much guidance.

Author: @emchap the goal of this being a reference to a single source of truth, which should be our documentation, tells me that if that doc isn't good enough the team in charge of it needs to improve that doc.

I don't disagree, there's a LOT of information on that page about editing your manifest, but I wholeheartedly think this should be a reference to official documentation and documentation should be updated by the team that owns that to ensure it's correct, accurate, and usable. A single source of truth that is our official documentation makes way more sense in terms of supporting customers and long-term scalability.

@amenne I'm curious about your thoughts on this since we chatted a bit about docs this week too and content in the docs versus navigation of the docs.

Reviewer: ```suggestion
Sourcegraph has a list of configurations available and recommended that need to be made to your manifest file to properly install Sourcegraph into your cluster. This includes how to customize the deployment using Kustomize—if you prefer to use Helm, please see the [Helm install instructions](https://docs.sourcegraph.com/admin/deploy/kubernetes/helm), instead. 

Please review the full configuration guide prior to proceeding: [Configure Sourcegraph with Kubernetes](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure). Pay particular attention to configuring a storage class for your cloud provider, configuring network access, and configuring external databases if you choose to use them.

Other sections of the configuration page should be reviewed, but are not required (for example, an external Jaeger instance is optional).
```

Get started by following these steps to fork the [`deploy-sourcegraph`](https://github.com/sourcegraph/deploy-sourcegraph) repository: [Getting started](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure#getting-started)

- Configurations:

>   - [License Key](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure#add-license-key)

Reviewer: ```suggestion
 ```
Same reason as above.

Author: Going to wait on this based on my [question above about that response](https://github.com/sourcegraph/customer-training/pull/5#discussion_r883038670)

(more so a note to myself on why I left the convo open, haha)

  - [Network Access](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure#configure-network-access)
  - [External PostgreSQL Database](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure#sourcegraph-databases)
  - [Scaling for services](https://docs.sourcegraph.com/admin/deploy/kubernetes/scale#tuning-replica-counts-for-horizontal-scalability)
  - [Cluster role administrator access](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)

### Install Sourcegraph
With your manifest file configured, ensure that you can [access your cluster](https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster/) with `kubectl` and `kubectl` is pointed to the proper context.

- Follow the [Direct installation](https://docs.sourcegraph.com/admin/deploy/kubernetes#direct-installation) steps to apply your manifest to your cluster
- Once your deployment is completed, you can access Sourcegraph via the [ingress](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure#ingress-controller-recommended) you configured in your manifest

## Next steps

> Congratulations, Sourcegraph is now deployed and you can begin to configure your instance. We recommend you review the [Configurating Sourcegraph](https://docs.sourcegraph.com/admin/config) documentation next to [add a code host](https://docs.sourcegraph.com/admin/external_service), [configure SSO](https://docs.sourcegraph.com/admin/auth), and more.

Reviewer: ```suggestion
Congratulations, Sourcegraph is now deployed and you can begin to configure your instance. We recommend you review the [Configurating Sourcegraph](https://docs.sourcegraph.com/admin/config) documentation next to [add a code host](https://docs.sourcegraph.com/admin/external_service), [configure authN and authZ](https://docs.sourcegraph.com/admin/auth), and more.
```
May want to consider linking to https://docs.sourcegraph.com/admin/config/authorization_and_authentication instead, since it's oriented more towards "if you want x do y" than the main auth docs.

Author: Before I accept this @emchap, can you add the suggestion with the other auth piece? That makes sense to me if that's used and referenced more often than the main auth documentation. I think you're saying to do it this way?

```suggestion
Congratulations, Sourcegraph is now deployed and you can begin to configure your instance. We recommend you review the [Configurating Sourcegraph](https://docs.sourcegraph.com/admin/config) documentation next to [add a code host](https://docs.sourcegraph.com/admin/external_service), [configure authN and authZ](https://docs.sourcegraph.com/admin/config/authorization_and_authentication), and more.
```
The original link has a lot of auth information though on it, is it really not useful and shouldn't be here? Maybe both?

Author: @emchap that looks great to me!
